lang/cs/Org.Apache.REEF.IMRU/OnREEF/IMRUTasks/UpdateTaskHost.cs

```diff
@@ -104,19 +113,53 @@ public byte[] Call(byte[] memento)
                 }
             }
 
-            MapInputWithControlMessage<TMapInput> stopMessage =
+            if (!_break)
+            {
+                MapInputWithControlMessage<TMapInput> stopMessage =
                     new MapInputWithControlMessage<TMapInput>(MapControlMessage.Stop);
-            _dataAndControlMessageSender.Send(stopMessage);
+                _dataAndControlMessageSender.Send(stopMessage);
+            }
 
             _resultHandler.Dispose();
+            _stopped = true;
             return null;
         }
 
         /// <summary>
+        /// Received close event. If it is closed command sent from driver, set _break to true to stop the iteration in Call method
+        /// After _enforceCloseTimeout, if the task is still not closed, throw IMRUTaskSystemException to enforce stop. 
+        /// </summary>
+        /// <param name="closeEvent"></param>
+        public void OnNext(ICloseEvent closeEvent)
+        {
+            if (closeEvent.Value.IsPresent() && Encoding.UTF8.GetString(closeEvent.Value.Value).Equals(TaskManager.CloseTaskByDriver))
+            {
+                Logger.Log(Level.Info, "The task received close event with message: {0}.", Encoding.UTF8.GetString(closeEvent.Value.Value));
+                _break = true;
+            }
+            Thread.Sleep(TimeSpan.FromMilliseconds(_enforceCloseTimeoutMilliseconds));
```

Comment:
That's right. 

But still we need a backup plan because we may miss some other scenarios. Because in any case, the task must be stopped in order to submit a new one, no matter how it is stopped.